catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/DynamicSchemaResolver.java

```diff
@@ -353,6 +268,114 @@ public void addFields(Metacard metacard, SolrInputDocument solrInputDocument)
     solrInputDocument.addField(SchemaFields.METACARD_TYPE_OBJECT_FIELD_NAME, metacardTypeBytes);
   }
 
+  private void addAttributeValues(Metacard metacard, SolrInputDocument solrInputDocument)
+      throws MetacardCreationException {
+    MetacardType schema = metacard.getMetacardType();
+
+    for (AttributeDescriptor ad : schema.getAttributeDescriptors()) {
+      if (hasCorrespondingAttributeWithValues(metacard, ad)) {
+        List<Serializable> attributeValues = metacard.getAttribute(ad.getName()).getValues();
+        AttributeFormat format = ad.getType().getAttributeFormat();
+        String formatIndexName = ad.getName() + getFieldSuffix(format);
+
+        attributeValues =
+            tryAddingAttributeValuesBasedOnFormat(
+                solrInputDocument, ad, attributeValues, format, formatIndexName);
+
+        addSortKeyField(solrInputDocument, attributeValues, format, formatIndexName);
+
+        // If it hasn't already been added.
+        if (solrInputDocument.getFieldValue(formatIndexName) == null) {
+          solrInputDocument.addField(formatIndexName, attributeValues);
+        } else {
+          LOGGER.trace("Skipping adding field already found on document ({})", formatIndexName);
+        }
+      }
+    }
+  }
+
+  private List<Serializable> tryAddingAttributeValuesBasedOnFormat(
+      SolrInputDocument solrInputDocument,
+      AttributeDescriptor ad,
+      List<Serializable> attributeValues,
+      AttributeFormat format,
+      String formatIndexName)
+      throws MetacardCreationException {
+    if (AttributeFormat.XML.equals(format)
+        && solrInputDocument.getFieldValue(
+                formatIndexName + getSpecialIndexSuffix(AttributeFormat.STRING))
+            == null) {
+      List<String> parsedTexts = parseTextFrom(attributeValues);
+
+      // parsedTexts => *_txt_tokenized
+      String specialStringIndexName =
+          ad.getName()
+              + getFieldSuffix(AttributeFormat.STRING)
+              + getSpecialIndexSuffix(AttributeFormat.STRING);
+      solrInputDocument.addField(specialStringIndexName, parsedTexts);
+    } else if (AttributeFormat.STRING.equals(format)
+        && solrInputDocument.getFieldValue(ad.getName() + getFieldSuffix(AttributeFormat.STRING))
+            == null) {
+      List<Serializable> truncatedValues =
+          attributeValues
+              .stream()
+              .map(
+                  value ->
+                      value != null ? truncateAsUTF8(value.toString(), TOKEN_MAXIMUM_BYTES) : value)
+              .collect(Collectors.toList());
```

Comment:
❓  Since (I think) these are a collection of values for a single attribute, should we filter out `null` values before collecting to a list?

Reply:
I was just moving code around to simplify things and make Sonarqube happy. I'd rather avoid changing anything functionally 😄